File: src/StructTypes.java
import java.util.*;

class StructTypes
{
	private Hashtable<String, StructDef> definedTypes = new Hashtable<String, StructDef>();

	public static class StructDef
	{
		private String name;
		private Hashtable<String, Type> types = new Hashtable<String, Type>();

		private StructDef(String name)
		{
			this.name = name;
		}

		public String getName()
		{
			return name;
		}

		public boolean hasField(String f)
		{
			return types.containsKey(f);
		}

		public void addField(String f, Type t)
		{
			types.put(f, t);
		}
	}

	public boolean isDefined(String sym)
	{
		return definedTypes.containsKey(sym);
	}

	public StructDef define(String name)
	{
		StructDef sdef = new StructDef(name);
		this.definedTypes.put(name, sdef);
		return sdef;
	}
}

File: csharp/CSC431/StructTypes.cs
using System;
using System.Collections.Generic;
using System.Text;


public class StructTypes
{
    private Dictionary<String, StructDef> definedTypes = new Dictionary<String, StructDef>();

	public class StructDef
	{
		private String name;
        private Dictionary<String, Type> types = new Dictionary<String, Type>();

		internal StructDef(String name)
		{
			this.name = name;
		}

		public String getName()
		{
			return name;
		}

		public bool hasField(String f)
		{
			return types.ContainsKey(f);
		}

		public void addField(String f, Type t)
		{
			types.Add(f, t);
		}

		public Type getFieldType(String f)
		{
			return this.types[f];
		}
	}

	public bool isDefined(String sym)
	{
		return definedTypes.ContainsKey(sym);
	}

	public StructDef define(String name)
	{
		StructDef sdef = new StructDef(name);
		this.definedTypes.Add(name, sdef);
		return sdef;
	}

	public StructDef get(String name)
	{
		return this.definedTypes[name];
	}
}

